src/jdk.compiler/share/classes/com/sun/tools/javac/tree/JCTree.java

```diff
@@ -641,6 +641,8 @@ public static class JCPackageDecl extends JCTree implements PackageTree {
         /** The tree representing the package clause. */
         public JCExpression pid;
         public PackageSymbol packge;
+        /** Position of closing semicolon, optional. */
+        public int endPos = Position.NOPOS;
```

Comment:
Your code seems consistent with what we already do with `JCBody`, `JCSwicthExpression` and `JCSwitch`. There's nothing wrong with this approach, although if end positions are enabled, this might be a little wasteful, as we'll store end positions both in the table and in the tree... a possible alternative would have been to construct a "minimal" end pos table which only contains positions for some "important" declarations (even when end positions are switched off) since these seems necessary for the compiler to work correctly. That would allow all code to flow through the end pos table and maybe simplify the implementation of the `TreeInfo` method (as well as `storeEnd` in the parser). But I'll leave that option for you to consider.